src/java.base/share/classes/sun/security/util/PBEUtil.java

```diff
@@ -67,6 +67,51 @@ public static final class PBES2Params {
         private byte[] salt;
         private IvParameterSpec ivSpec;
 
+        /*
+         * Initialize a PBES2Params instance. May generate random salt and
+         * IV if not passed and the operation is encryption. If initialization
+         * fails, values are reset. Used by PBES2Params and P11PBECipher
+         * (SunPKCS11).
+         */
+        public void initialize(int blkSize, int opmode, int iCount, byte[] salt,
+                AlgorithmParameterSpec params, SecureRandom random)
+                throws InvalidAlgorithmParameterException {
+            try {
+                boolean doEncrypt = ((opmode == Cipher.ENCRYPT_MODE) ||
+                        (opmode == Cipher.WRAP_MODE));
+                if (params instanceof PBEParameterSpec pbeParams) {
+                    params = pbeParams.getParameterSpec();
+                }
+                if (params instanceof IvParameterSpec iv) {
+                    this.ivSpec = iv;
+                } else if (params == null && doEncrypt) {
+                    byte[] ivBytes = new byte[blkSize];
+                    random.nextBytes(ivBytes);
+                    this.ivSpec = new IvParameterSpec(ivBytes);
+                } else {
+                    throw new InvalidAlgorithmParameterException("Wrong " +
+                            "parameter type: IvParameterSpec " +
+                            (doEncrypt ? "or null " : "") + "expected");
+                }
+                this.iCount = iCount;
+                if (salt == null) {
+                    if (doEncrypt) {
+                        salt = new byte[DEFAULT_SALT_LENGTH];
+                        random.nextBytes(salt);
+                    } else {
+                        throw new InvalidAlgorithmParameterException("Salt " +
+                                "needed for decryption");
+                    }
+                }
```

Comment:
Isn't there also default value for iteration count? 'params' can be PBEParameterSpec (line 82) but its salt and iteration count values aren't used comparing to the IvParameterSpec inside. Seems a bit inconsistent?

Reply:
@valeriepeng: I agree, `DEFAULT_ITERATIONS` should be used here and only here, so we consistently initialize any defaults in a single place. We'll update that.

Reply:
There are cases in which _salt_ and _iteration count_ come from the _key_ (`javax.crypto.interfaces.PBEKey`), so `PBES2Params.initialize()` shouldn't try to extract them from `params` (when it is `PBEParameterSpec`). To solve the inconsistency, and in line with your other comment, let's allow `params` to only be `IvParameterSpec`. But let's still receive it as `AlgorithmParameterSpec`, so we can check its type here (and make it simpler for callers), where only `IvParameterSpec` is allowed (plus `null` when encrypting).

Reply:
Sure, sounds reasonable.

Reply:
Both changes (passing a IvParameterSpec instance and consolidating defaults for iteration count in a single place) look good to me.